gene/database/dynamodb.py

```diff
@@ -361,22 +365,120 @@ def update_record(self, concept_id: str, field: str, new_value: Any,
         :param str new_value: new value
         :param str item_type: record type, one of {'identity', 'merger'}
+        :raise DatabaseWriteException: if attempting to update non-existent record
         """
+        if field == "merge_ref":
+            new_value = new_value.lower()
+        label_and_type = f"{concept_id.lower()}##{item_type}"
         key = {
-            'label_and_type': f'{concept_id.lower()}##{item_type}',
-            'concept_id': concept_id
+            "label_and_type": label_and_type,
+            "concept_id": concept_id
         }
         update_expression = f"set {field}=:r"
         update_values = {':r': new_value}
+        condition_expression = "attribute_exists(label_and_type)"
         try:
             self.genes.update_item(Key=key,
                                    UpdateExpression=update_expression,
-                                   ExpressionAttributeValues=update_values)
+                                   ExpressionAttributeValues=update_values,
+                                   ConditionExpression=condition_expression)
         except ClientError as e:
-            logger.error(f"boto3 client error in `database.update_record()`: "
-                         f"{e.response['Error']['Message']}")
+            code = e.response.get("Error", {}).get("Code")
+            if code == "ConditionalCheckFailedException":
+                raise DatabaseWriteException(
+                    f"No such record exists for keys {label_and_type}, {concept_id}"
+                )
+            else:
+                logger.error(f"boto3 client error in `database.update_record()`: "
+                             f"{e.response['Error']['Message']}")
+
+    def delete_normalized_concepts(self) -> None:
+        """Remove merged records from the database. Use when performing a new update
+        of normalized data.
+
+        :raise DatabaseReadException: if DB client requires separate read calls and
+            encounters a failure in the process
+        :raise DatabaseWriteException: if deletion call fails
+        """
+        while True:
+            with self.genes.batch_writer(
+                overwrite_by_pkeys=["label_and_type", "concept_id"]
+            ) as batch:
+                try:
+                    response = self.genes.query(
+                        IndexName="item_type_index",
+                        KeyConditionExpression=Key("item_type").eq("merger"),
+                    )
+                except ClientError as e:
+                    raise DatabaseReadException(e)
+                records = response["Items"]
+                if not records:
+                    break
+                for record in records:
+                    batch.delete_item(Key={
+                        "label_and_type": record["label_and_type"],
+                        "concept_id": record["concept_id"]
+                    })
+
+    def delete_source(self, src_name: SourceName) -> None:
```

Comment:
Should we consider using a transaction?

Reply:
I didn't know that was a thing for dynamo! Going to make a new issue